esphome/components/waveshare_epaper/waveshare_epaper.cpp

```diff
@@ -1507,6 +1507,326 @@ void WaveshareEPaper2P9InV2R2::set_full_update_every(uint32_t full_update_every)
+                                                         // 8 pixel, 1 for black and 1 for red)
+      for (y = 0; y < height_px; y++) {
+        for (x = 0; x < width_b; x++) {
+          uint16_t segment_position = x + y * width_b;  // linear position inside the segment in bytes
+          uint16_t global_position = seg_y * height_px * this->get_width_internal() / 8 +
+                                     y * this->get_width_internal() / 8 + seg_x * width_b + x;
+          segment[segment_position] = buffer_[global_position];                                    // copy black data
+          segment[segment_position + segment_size] = buffer_[global_position + buffer_half_size];  // copy red data
+        }
+      }
+      // now calculate a CRC16_checksum and compare it against the stored value.
+      // ESP_LOGD(TAG, "Calculating a CRC");
+      uint16_t segment_crc = crc16(segment, segment_size * 2, 65535U, 40961U, false, false);
+      if (full_sync) {
+        // no need to compare, we're in the first run, just place it. This is called by full refresh only
+        this->checksums_[seg_x + seg_y * seg_x_] = segment_crc;
+      } else {
+        // Partial Update, compare checksums while replacing and record the X and Y block position of the top left and
+        // bottom right corner of the changed elements. Afterwards, we can partially update only the segment that has
+        // been altered.
+        bool changed = this->checksums_[seg_x + seg_y * seg_x_] != segment_crc;
+        this->checksums_[seg_x + seg_y * seg_x_] = segment_crc;
+        if (changed && !found_change) {
+          found_change = true;
+          // We need to span the x segment, with the lowest segment found making the first segment and the highest
+          // segment found the last segment.
+          if (seg_x < this->first_segment_x_)
+            this->first_segment_x_ = seg_x;
+          if (seg_x > this->last_segment_x_)
+            this->last_segment_x_ = seg_x;
+          this->first_segment_y_ = seg_y;
+        } else if (changed) {
+          // We need to span the x segment, with the lowest segment found making the first segment and the highest
+          // segment found the last segment.
+          if (seg_x < this->first_segment_x_)
+            this->first_segment_x_ = seg_x;
+          if (seg_x > this->last_segment_x_)
+            this->last_segment_x_ = seg_x;
+          // Segment changed but we already found a change before, so set it as last segment
+          this->last_segment_y_ = seg_y;
+        } else {
+          // do nothing, segment didn't change.
+        }
+      }
+      delete[] segment;  // Delete the heap element again, this is not java... >.<
+    }
+  }
+  if (found_change) {
+    ESP_LOGD(TAG, "CRC Calculation finished. Found changes: %02d:%02d to %02d:%02d", first_segment_x_, first_segment_y_,
+             last_segment_x_, last_segment_y_);
```

Comment:
Please ensure all member variable access and function calls are prefixed with `this->`

```suggestion
    ESP_LOGD(TAG, "CRC Calculation finished. Found changes: %02d:%02d to %02d:%02d", this->first_segment_x_, this->first_segment_y_,
             this->last_segment_x_, this->last_segment_y_);
```